File: Algorithms/List/206. Reverse Linked List.py
#iterative
class Solution(object):
    def reverseList(self, head):
        if(not head):
            return head
        p = head.next
        head.next = None
        q = head
        while(p):
            temp = p.next
            p.next = q
            q = p
            p = temp
        return q

#recursive
class Solution(object):
    def reverseList(self, head):
        def rev(node,pre=None):
            if(not node):
                return pre
            temp = node.next
            node.next = pre
            return rev(temp, node)
        return rev(head)

File: Algorithms/List/206. Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
//iterative
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
		ListNode* cur = head;
		ListNode* prev = nullptr;
		while(cur != nullptr)
		{
			ListNode* temp = cur->next;
			cur->next = prev;
			prev = cur;
			cur = temp;
		}
		return prev;
    }
};

//recursion
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
		if(!head || !head->next) return head;
		ListNode* p = reverseList(head->next);
		head->next->next = head;
		head->next = nullptr;
		return p;
	}
};
